### ISSUE TITLE
Return value and `after` stage errors

### ISSUE THREAD

**Issue by toddbaert** [open-feature/spec#174]
As pointed out [here](https://github.com/open-feature/dotnet-sdk/pull/109), the spec doesn't dictate whether the `evaluated flag value`, or the `default value` should be returned in the case of an error in the AFTER stage hooks ([4.4.7](https://github.com/open-feature/spec/blob/main/specification/sections/04-hooks.md#requirement-447) does say: `If an error occurs in the before hooks, the default value MUST be returned.`).

I think since "after" are really about side-effects, like telemetry, etc, we probably should return the evaluated value. I'd like other's thoughts.

cc @justinabrahms @skyerus @moredip @beeme1mr @benjiro @kinyoklion @josecolella

**Comment by beeme1mr** [comment]
Yeah, good catch. Throwing in an after hook is how we're currently performing validation in the playground.

https://github.com/open-feature/playground/blob/main/packages/app/src/app/hex-color/hex-color.service.ts#L13-L25

I would like to maintain the ability to perform validation via a hook but it could be implemented in a different way.

**Comment by toddbaert** [comment]
@beeme1mr I forgot about the "validating in the after by throwing" use case... Since we don't allow mutation of the returned value, and this is what most SDKs have done. Maybe we should keep the current behavior and just specify it rather than making a surprise change.

**Comment by justinabrahms** [comment]
I kinda don't like the hook for this. It doesn't feel like a hook should be the mechanism of data validation to me (though perhaps we need something like that).

```js
const hexColorValue = await this.client.getStringValue('hex-color', '000000', undefined, {
    hooks: [
      {
        after: (hookContext, evaluationDetails: EvaluationDetails<string>) => {
          if (!is_hex(evaluationDetails.value)) {
            throw new Error(`Invalid hex value: ${evaluationDetails.value}`);
          }
        },
      },
    ],
});
return `#${hexColorValue}`;
```

would become

```js
const default = '000';
const hexColorValue = await this.client.getStringValue('hex-color', default);
if !is_hex(hexColorValue) {
  return default
}
return hexColorValue
```
(I don't like that default variable though...)

perhaps most ideally:
```js
return await this.client.getStringValue('hex-color', default, undefined, { valueValidator: is_hex });
```

and the "didn't pass validation" could be on the detail object.

I suppose my counterargument (to myself) would be this, wherein `ValueValidator` is a type of hook in userland.
```js
return await this.client.getStringValue('hex-color', default, undefined, { hooks: ValueValidator(is_hex) });
```

Spitballin'.

**Comment by beeme1mr** [comment]
I think hooks work well for validation but perhaps a dedicated stage would make the behavior more clear. I'll think about this and see if I can come up with a proposal.

**Comment by kinyoklion** [comment]
My initial reflex would be that I don't like the idea of an after hook changing the result. Though maybe more generally speaking I would prefer a differentiation between hooks which can affect results and those that cannot (or maybe meta-data for the hook?). So if my telemetry provider makes some mistake, then I don't start getting default values for all my variations.

I do think that having some validation mechanism may be nice. It would allow making some middleware type library that might let you do something like:
```
{
	myInt: {type: "integer", min: 0, max 255},
	myString: type: "string", matches: /[0-9A-Fa-f]{6}/g } 
}
```
(Though in a few languages this might be better served using the type system of the language.)

**Comment by justinabrahms** [comment]
@beeme1mr The question I'm thinking is: How do we not do that at all, and instead give userland the tools to do it?

I think that would be adding support for an explicit "validation" stage within the hook lifecycle and then letting folks figure out if/how they'll register their rules.

**Comment by beeme1mr** [comment]
I probably shouldn't have said "dedicated validation system". That sounds more complex than I had in mind. Basically, I think we could add support for calling a validate method during flag evaluation. That could be a new stage in a hook or a new flag evaluation option.

**Comment by justinabrahms** [comment]
@toddbaert That would be a win in my book. :)

**Comment by kinyoklion** [comment]
> I do agree with @kinyoklion that it would be nice if there was a way for application authors and hook authors to GUARANTEE that a hook wouldn't cause a default if it's only responsible for some side-effect (like tracking). I think that's _sort of_ handled by hook authors putting their actions in a try/catch and swallowing errors, but I think it's possible people will forget to do this.

I was somewhat thinking the opposite here, it would have been nice if after didn't mutate.

That said, it could be handled as metadata on a hook instead. Existing hooks could be mutating, and a `readonly` or some other attribute or metadata could be added to indicate strictly readonly hooks.

**Comment by toddbaert** [comment]
> I was somewhat thinking the opposite here, it would have been nice if after didn't mutate and a validate did.

Are you suggesting that we change things such that throwing in an `after` wouldn't change the evaluation result, while such things in `validate` would? And how do you feel about actually changing the returned result in validate, would that be supported? Last question: do you think semantics of `before` should change in any similar respects?

Wherever we land, I'd like to maintain the principle of least astonishment as much as possible, and err on the side of simplicity.

**Comment by kinyoklion** [comment]
> > I was somewhat thinking the opposite here, it would have been nice if after didn't mutate and a validate did.
> 
> Are you suggesting that we change things such that throwing in an `after` wouldn't change the evaluation result, while such things in `validate` would? And how do you feel about actually changing the returned result in validate, would that be supported? Last question: do you think semantics of `before` should change in any similar respects?
> 
> Wherever we land, I'd like to maintain the principle of least astonishment as much as possible, and err on the side of simplicity.

As a consumer of the API I don't like the idea of my analytics or debugging hooks having the same capabilities as my context manipulation hooks or validation type hooks. So I am just trying to consider a world where I have some level of control over that. 

To that end, potentially validate versus after doesn't give the consumer of the API control really. Because all the functionality is available in every hook, because hooks are an interface with many functions (versus specific hook types). 

Given that interface as a pre-condition, then I think that it makes sense for the entire hook to be either readonly or mutating. Then the stage of those mutations doesn't really matter.

Second I want to know that the mutation happened. Something should indicate that the value was modified and ideally the origin. This largely for provider authors and vendors, so when evaluations aren't what people expect, the cause can be quickly identified.

**Comment by beeme1mr** [comment]
I think validation is the only use case where we would want to mutate the response. Also, a validator should only run during a flag evaluation. Perhaps hooks should behave as @kinyoklion describes, and an optional `validate` argument could be added.

```typescript
const hexValidator = (value) => {
	if (!value.match(/[0-9A-Fa-f]{6}/g)) {
		throw new ParseError("not a valid hex");
	}
}

await this.client.getStringValue(
	'hex-color', 
	'000000', 
	undefined,
	// Note: Only a single validator can be passed
	{ validate: hexValidator });
```

This has a few advantages:

- Hooks can no longer affect the return value.
- It's impossible to accidentally register a validation hook at the global or client level.
- Easy to use and understand as a developer.

What do you think?

**Comment by toddbaert** [comment]
> I think validation is the only use case where we would want to mutate the response. Also, a validator should only run during a flag evaluation. Perhaps hooks should behave as @kinyoklion describes, and an optional `validate` argument could be added.
> 
> ```ts
> const hexValidator = (value) => {
> 	if (!value.match(/[0-9A-Fa-f]{6}/g)) {
> 		throw new ParseError("not a valid hex");
> 	}
> }
> 
> await this.client.getStringValue(
> 	'hex-color', 
> 	'000000', 
> 	undefined,
> 	// Note: Only a single validator can be passed
> 	{ validate: hexValidator });
> ```
> 
> This has a few advantages:
> 
> * Hooks can no longer affect the return value.
> * It's impossible to accidentally register a validation hook at the global or client level.
> * Easy to use and understand as a developer.
> 
> What do you think?

I'm OK with some of the ideas here, but I think I'd prefer to add validate as another hook stage (as @kinyoklion seems to be indicating) not a new concept.
